---
title: "Analysis Report 1: Influence of sex on the diversity of hand surface bacteria."
author: "Allison Bogisich"
date: "October 31, 2017"
output: github_document
bibliography: references.bib
csl: bioinformatics.csl
---

# Introduction. 

Bacteria thrive not only within, but on almost every surface off the human body. Especially of interest is the microbiome maintained on the largest human organ: the skin. As such, it harbors one of the largest bacterial communities, and has vast potential to have significant impacts on one's health (@dethlefsen2007ecological, @turnbaugh2007human). Skin is a complex habitat with both local and regional variations in not only amount of environmental exposure, but also the cellular make-up of the particular dermal variety (@fredricks2001microbial). The great majority of these bacteria are not merely passive nor transient occupiers of the skin's surface, but instead are uniquely adapted for the particular niches made available in differing regions of the skin (@fierer2008influence). Regions of frequent skin shedding, antimicrobial host defenses, and variance in the amount of exposure to soaps or detergents when cleansing, exposure to UV radiation, and availability of moisture sources are all factors that help to shape the microbial communities which can persist (@roth1988microbial, @cogen2008skin). 

The bacteria on skin surfaces appears to be highly diverse, however much of the diversity has not been extensively mapped. In addition to having difficulty in identifying the sheer number of bacterial species in a heterogeneous sample, certain culturing methods have shown that some taxa are specific to the skin environment and are difficult to grow in mono culture (@chiller2001skin). Previous research has also observed consistent amounts of variability between individuals in the composition of their skin microbiomes (@gao2007molecular, @grice2008diversity). Many questions still remain unanswered about the type of environmental and genetic factors that could be contributing to-or even driving-this variability. 

With human DNA, it is highly conventional and relatively easy to determine biological gender based on chromosomal signature alone. Is it then entirely possible that our microbiomes carry a gender signature as well, since they are so intrinsically tied to us. In mice, stomach acid analyses have shown gender-specific differences as well as paralleled effects given gonad removal, providing support for mechanisms mediating sex differences in microbiota composition (@mehrabian2016sex). Determining how much microbiota community structure is shaped by both environmental and particularly genetic factors, and which particular aspects are linked to each respective factor could have significant implications for the emerging field of bio-forensics.

Only within the past decade or so has the technical capability to sequence complex biological samples  been able to take off to such an astounding extent. Recent developments in phylogenetic community analysis (@lozupone2005unifrac) with high-throughput pyrosequencing methods (@hamady2008error) have established the criteria for not only sequencing microbiome samples, but the very foundations of bacterial forensics. Fierer and his team have conducted a series of studies that demonstrate the strengths and weaknesses of the field (@fierer2010forensic). They found that skin-associated bacterial communities are surprisingly diverse, and that they are independently stable enough to assist traditional forensic evaluations.

Conventional methods of obtaining forensic results from human DNA require sufficient amounts of blood, tissue, semen, or saliva on an object. However, it is often difficult to obtain a large (and uncontaminated) enough sample to sequence. In order to boost standard results, recovering bacterial DNA from touched surfaces may be far easier, especially for identifying objects where clear fingerprints can't be obtained (e.g. fabrics, smudged surfaces, highly textured surfaces) (@fierer2008influence). Given how abundant bacterial cells are on the surface of skin and on shed epidermal cells (@fredricks2001microbial) and how highly personalized bacterial communities are implicates that more research is needed. In this post-analysis of Fierer's research, I now report the degree of sexual dimorphism that exists between male versus female subject's skin microbiomes from the aforementioned project. If a gender signature does exist in skin microbiomes, then I would predict that there will be marked differencies in the bacterium making up sebum communities in each respective gender sampled.


# Methods

This post-experimental report is based on metadata output files from a BLAST search of trimmed and quality checked fasta files from the Fierer team study. The BLAST search matched swabbed and sequenced sebum samples against the GenBank database. Data used was from from NCBI Sequence Read Archive study number ERP022657, of which a summary of the information is available [here](https://www.ncbi.nlm.nih.gov/Traces/study/?WebEnv=NCID_1_128047291_130.14.22.33_5555_1505945515_1626731749_0MetA0_S_HStore&query_key=5). The metadata output process and the resulting files from this study can be found in a Github repository in a `data/metadata` directory available [here] https://github.com/2017-usfca-cs-640/ABogisich-analysis-report-01-amplicons. Using bioinformatic applications in R and R Studio, the metadata output from this repository were reorganized into subsets and transfigured into easily view-able figures to analyze and compare the diversity and potential sexual dimorphism of microbial fauna present on female and male subjects.



## Sample origin and sequencing

Fierer and his team swabbed for skin-associated bacteria from different epidermal regions for each of their three studies. For the keyboard study, three participants were swabbed on the ventral surface of the distal joint of each fingertip. All individuals were healthy 20-35 year olds that had not taken antibiotics at least six months prior to swabbing. In the "storage" study, autoclaved cotton-tipped swabs moistened with sterile solution were again used, this time to sample the right axilary (armpit) skin surface sixteen times, from two healthy adults. Lastly, in the computer mouse study nine healthy adults were recruited (four female and five male, all 20-35 years of age) who worked in the same University of Colorado building. Using the swabbing technique outlined above, the entirety of the exposed surface of their dominant hand's palm (used to control the computer mice) were swabbed. Palm surfaces were sampled at midday and the participants had been following typical hand hygiene practices prior to sampling. Swabs were stored at -80 degrees before DNA extraction. The microbial communities on these participants were compared to a previously compiled database from 270 other hands sampled by Fierer and collaborators (@fierer2008influence; @costello2009bacterial). The 270 hands in the database were from left and right palm surfaces belonging to an equal proportion of both healthy male and female volunteers, aged 18-40 years old.

Post DNA extraction from the samples, sequences were processed and analyzed using pyrosequencing procedures akin to Fierer's work in 2008. Sequences were removed if shorter than 200 bp or larger than 300 in length, had a quality score lower than 25, had ambiguous characters or uncorrectable barcodes, or if it did not contain the primer sequence.


## Computational

The program R version 3.4.2 in conjunction with R Studio version 1.1.383 interface for Windows was used. Several general use packages from the R library were downloaded in order to utilize more specific exploratory data analysis functions. See listing under the code chunk "r load-libraries" for more detailed installation information. References can be found in the references.bib directory, as well as at the bottom of this report. Citations were organized and inserted using citr, accessible [here] https://CRAN.R-project.org/package=citr.

Primary packages used to clean and analyze the data were installed via bioconductor and devtools. The DADA2 package [@callahan2016] was installed for use in conducting fast and accurate sample inferences from amplicon data with single-nucleotide resolution. Seqinr was used to export fastas of our final de-noised variants. Devtools was installed prior to downloading of mctoolsr, which stands for “microbial community analysis tools in R”. Created by Fierer, the intention of the package is to make it faster to manipulate data and perform certain analyses common to microbial community studies, while allowing us to easily access data using established functions (Leff, 2017). Phyloseq [@mcmurdie2013] was the final package installed for my pipeline in order to visualize and quality check my results using tools `plot_bar`, `plot_richness`, `ploterrors` and `plotQualityProfile`.

Raw data was sorted for order and samples were extracted with their appropriate naming system. Base paths and full paths were set for the input and outputs of sample extraction. Quality checks were run prior to any data visualization to ensure that the lengths of the sequences being used were of proper length for obtaining quality scores above 30, which required filtering and trimming all sequences. Error models were also created to analyze each sample individually, and then visualized by plotting. Duplicated sequences and chimeras were then removed from the data set, so as not to compound error.
`Seqinr` was used next to export the cleaned, trimmed, filtered, de-noised sequence variants so that I could build tables for use in making subset objects of data for desired visual graphics.


# Results

In addition to a minimum of 3-4 figures/tables (and associated captions), you should include sufficient text in this section to describe what your findings were. Remember that in the results section you just describe what you found, but you don't interpret it - that happens in the discussion.

```{r load-libraries, message = FALSE}
# Be sure to install these packages before running this script
# They can be installed either with the intall.packages() function
# or with the 'Packages' pane in RStudio

# load general-use packages
library("dplyr")
library("tidyr")
library("knitr")
library("ggplot2")

# this package allows for the easy inclusion of literature citations in our Rmd
# more info here: https://github.com/crsh/citr
# and here:
# http://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html
library("citr")

# These are the primary packages well use to clean and analyze the data
# this package needs to be installed from bioconductor -- it's not on CRAN
# see info here: https://benjjneb.github.io/dada2/dada-installation.html
library("dada2")

# This to export a fasta of our final de-noised sequence variants
library("seqinr")

# To install this you have to install from GitHub
# See more info here: https://github.com/leffj/mctoolsr
# run this -- install.packages("devtools")
# and then this -- devtools::install_github("leffj/mctoolsr")
library("mctoolsr")

# And this to visualize our results
# it also needs to be installed from bioconductor
library("phyloseq")
```

```{r extract-sample-and-file-names}
# NOTE: Much of the following follows the DADA2 tutorials available here:
# https://benjjneb.github.io/dada2/tutorial.html
# Accessed October 19, 2017

# set the base path for our input data files
path <- "data/raw_data"

# Sort ensures samples are in order
filenames_forward_reads <- sort(list.files(path, pattern = ".fastq"))

# Extract sample names, assuming filenames have format: SAMPLENAME.fastq
sample_names <- sapply(strsplit(filenames_forward_reads, "\\."), `[`, 1)

# Specify the full path to each of the filenames_forward_reads
filenames_forward_reads <- file.path(path, filenames_forward_reads)
```

```{r check-quality-plots}
# Plots the quality profiles of all twenty samples
plotQualityProfile(filenames_forward_reads[1:20])
```

We can see from the quality profiles that most reads tend to get pretty bad in quality after around 200 bases. Therefore, we decided to set a maximum acceptable sequence length of 225 bases.

```{r filter-reads}
# Place filtered files in filtered/ subdirectory
# note this will fail if the directory doesn't exist
filter_path <- file.path("output", "filtered")
filtered_reads_path <- file.path(filter_path,
                                 paste0(sample_names,
                                        "_filt.fastq.gz"))

# See ?filterAndTrim for details on the parameters
# See here for adjustments for 454 data:
# https://benjjneb.github.io/dada2/
#     faq.html#can-i-use-dada2-with-my-454-or-ion-torrent-data
filtered_output <- filterAndTrim(fwd = filenames_forward_reads,
                                 filt = filtered_reads_path,
                                 maxLen = 225,
                                 maxN = 0, # discard any seqs with Ns
                                 maxEE = 3, # allow w/ up to 3 expected errors
                                 truncQ = 2, # cut off if quality gets this low
                                 rm.phix = TRUE,
                                 compress = TRUE,
                                 multithread = FALSE)
```

```{r filtered-read-counts-table}
# produce nicely-formatted markdown table of read counts
# before/after trimming
kable(filtered_output,
      col.names = c("Reads In",
                    "Reads Out"))
```

```{r learn-errors}
# this build error models from each of the samples
errors_forward_reads <- learnErrors(filtered_reads_path,
                                    multithread = FALSE)
```

```{r visualize-errors-with-plots}
# quick check to see if error models match data
# (black lines match black points) and are generally decresing left to right
plotErrors(errors_forward_reads,
           nominalQ = TRUE)
```

```{r dereplicate-sequences}
# get rid of any duplicated sequences
dereplicated_forward_reads <- derepFastq(filtered_reads_path,
                                         verbose = TRUE)

# Name the derep-class objects by the sample names
names(dereplicated_forward_reads) <- sample_names
```

```{r run-dada}
# parameters adjusted based on recommendations for 454 data here:
# https://benjjneb.github.io/dada2/
#     faq.html#can-i-use-dada2-with-my-454-or-ion-torrent-data
dada_forward_reads <- dada(dereplicated_forward_reads,
                           err = errors_forward_reads,
                           HOMOPOLYMER_GAP_PENALTY = -1, # reduce penalty bc 454
                           BAND_SIZE = 32) # performs local alignments bc indels

# check dada results
dada_forward_reads
```

```{r make-sequence-table}
# produce the 'site by species matrix'
sequence_table <- makeSequenceTable(dada_forward_reads)
```

The output table has `r nrow(sequence_table)` rows (samples) and `r ncol(sequence_table)` columns (sequence variants). Notice how we can embed R code directly in our markdown text.

```{r histogram-of-sequence-lengths}
# Quick check to look at distribution of trimmed and denoised sequences
hist(nchar(getSequences(sequence_table)),
     main = "Histogram of fingal sequence variant lengths",
     xlab = "Sequence length in bp")
```

```{r remove-chimeras}
# Check for and remove chimeras
sequence_table_nochim <- removeBimeraDenovo(sequence_table,
                                            method = "consensus",
                                            multithread = FALSE,
                                            verbose = TRUE)

# What percent of our reads are non-chimeric?
non_chimeric_reads <- round(sum(sequence_table_nochim) / sum(sequence_table),
                            digits = 4) * 100
```

After removing chimeras, we were left with `r non_chimeric_reads`% of our cleaned reads.

```{r table-of-pipeline-read-counts}
# Build a table showing how many sequences remain at each step of the pipeline
get_n <- function(x) sum(getUniques(x)) # make a quick function
track <- cbind(filtered_output, # already has 2 columns
               sapply(dada_forward_reads, get_n),
               rowSums(sequence_table),
               rowSums(sequence_table_nochim))

# add nice meaningful column names
colnames(track) <- c("Input",
                     "Filtered",
                     "Denoised",
                     "Sequence Table",
                     "Non-chimeric")

# set the proper rownames
rownames(track) <- sample_names

# produce nice markdown table of progress through the pipeline
kable(track)
```

```{r assign-taxonomy}
# assigns taxonomy to each sequence variant based on a supplied training set
# made up of known sequences
taxa <- assignTaxonomy(sequence_table_nochim,
                       "data/training/rdp_train_set_16.fa.gz",
                       multithread = FALSE,
                       tryRC = TRUE) # also check with seq reverse compliments

# show the results of the taxonomy assignment
unname(taxa)
```

```{r extract-sequences-to-fasta}
# we want to export the cleaned, trimmed, filtered, de-noised sequence variants
# so that we can build tables - phylogenies can be built outside of R-
# but we will need a fasta file to do so. We keep the names of each sequence as the
# sequence itself (which is QUITE confusing), because that's how DADA2 labels
# it's columns (e.g. 'species')
# function taken from https://github.com/benjjneb/dada2/issues/88
export_taxa_table_and_seqs <- function(sequence_table_nochim,
                                       file_seqtab,
                                       file_seqs) {
  seqtab_t <- as.data.frame(t(sequence_table_nochim)) # transpose to data frame
  seqs <- row.names(seqtab_t) # extract rownames
  row.names(seqtab_t) <- seqs # set rownames to sequences
  outlist <- list(data_loaded = seqtab_t)
  mctoolsr::export_taxa_table(outlist, file_seqtab) # write out an OTU table
  seqs <- as.list(seqs)
  seqinr::write.fasta(seqs, row.names(seqtab_t), file_seqs) # write out fasta
}

# actually run the function, with the names of the files we want it to create
# and where to put them
export_taxa_table_and_seqs(sequence_table_nochim,
                           "output/sequence_variants_table.txt",
                           "output/sequence_variants_seqs.fa")
```


```{r read-in-metadata-and-create-phyloseq}
# Next we want to read in the metadata file so we can add that in too
# This is not a csv file, so we have to use a slightly different syntax
# here the `sep = "\t"` tells the function that the data are tab-delimited
# and the `stringsAsFactors = FALSE` tells it not to assume that things are
# categorical variables
metadata_in <- read.table(paste0("data/metadata/",
                                 "fierer_forensic_hand_mouse_SraRunTable.txt"),
                          sep = "\t",
                          header = TRUE,
                          stringsAsFactors = FALSE,
                          row.names = 6) # sets sample IDs to row names

# read in the phylogeny, which was created from the fasta exported above
# in Geneious by aligning the sequences with MAFFT and then building a
# Maximum-Likelihood tree with RAxML
tree_in <- read_tree("output/sequence_variants_MAFFT_RAxML.newick")

# Construct phyloseq object (straightforward from dada2 outputs)
phyloseq_obj <- phyloseq(otu_table(sequence_table_nochim,
                                   taxa_are_rows = FALSE), # sample-spp matrix
                         sample_data(metadata_in), # metadata for each sample
                         tax_table(taxa), # taxonomy for each sequence variant
                         phy_tree(tree_in)) # phylogeny from sequence variants
```


```{r alpha-diversity-skin}
# alpha diversity metrics, create subset
sebum_subset_obj <- subset_samples(phyloseq_obj,
                                   sex_s != "Not applicable")
# plot alpha diversity of sebum samples
plot_richness(sebum_subset_obj,
              x = "env_material_s",
              measures = c("Shannon", "Simpson"),
              color = "sex_s") +
  xlab("Sample Origin") +
  geom_jitter(width = 0.2) +
  theme_bw()

```
Figure 1: Alpha diversity measures of female and male samples, colored by gender.

The communities in this subset are rather under-sampled in terms of sequence depth in this data set. However, by using the Shannon and Simpson diversity indexes I have plotted the mean species diversity in sites or habitats at a more local/individual's skin scale. The majority of red data points -which indicate female sebum samples- are generally clustering more at the upper half of both diversity indexes, with a few outliers farther down the diversity scale. Conversely, the blue male data points are largely clustering in the lower half of the scale for alpha diversity in both indexes. 


```{r male-microbiome}
# construct male participant object
subset_male_obj <- subset_samples(phyloseq_obj,
                                  sex_s == "male")
```

```{r male-subset-plot-2}
# graph male microbiome relative abundance categorized by phyla
plot_bar(subset_male_obj,
         fill = "Phylum")
```
Figure 2: Relative abundance of bacterial phyla in male sebum samples.

For the five male sebum samples collected, the phylum making up the vast majority of all microbiomes is Proteobacteria. While the abundance does range from about 220 to about 820, it is always the phylum with the greatest representation for each of the men's invidual bacterial skin community. One of the samples (ERR1942291) exhibits a considerable abundance of sequences that matched for cyanobacteria and/or cholorplast DNA, while a smaller amount of that phylum is also found in sample ERR1942293. There were also small quantites of male skin bacteria that resulted as sequence matches for Actinobactera. The total levels of bacterial abundance found in each of the male samples were also not consistent, ranging anywhere from about 230 to near 850 at the highest.

```{r female-microbiome}
# construct female particpant object
subset_female_obj <- subset_samples(phyloseq_obj,
                                    sex_s == "female")
```

```{r female-subset-plot-3}
# graph female microbiome relative abundance categorized by phyla
plot_bar(subset_female_obj, 
         fill = "Phylum")
```
Figure 3: Relative abundance of bacterial phyla in female sebum samples.

Out of the five female sebum samples collected, the phylum with the greatest overall representation is Proteobacteria. Its abundance ranges from about 20 to about 510, however it was not found in each of the women's invidual bacterial skin communities nor did it always make up the largest phylum in their respective communites. The female samples appear to be much more diverse in the types and amounts of each phyla present. There were also noticeable quantites of female skin bacteria that resulted as sequence matches for Actinobactera. The total levels of bacterial abundance found in each of the female samples were also not consistent, ranging anywhere from about 40 to 550.

```{r male-microbiome-genus-plot-4}
# graph male microbiome relative abundance categorized by genus
plot_bar(subset_male_obj,
         fill = "Genus")
```
Figure 4: Relative abundance of bacterial genera in male sebum samples.

At the level of genera identification, it became much harder to identify levels of bacterial abundance. In all but one of the male samples, the majority of the DNA sequences did not have a known match in the database. In only one sample -ERR1942293- were the majority of the sequences matched to a genus, *Acinetobacter*.


```{r female-microbiome-genus-plot-5}
# graph female microbiome relative abundance categorized by genus
plot_bar(subset_female_obj,
         fill = "Genus")
```
Figure 5: Relative abundance of bacterial genera in female sebum samples.



# Discussion

Add around 2-3 pages interpreting your results and considering future directions one might take to analyze these data.

--It would be rather shocking if at the *phylum* level there were statistically significant differences between men and women. It’s just such a high level of taxonomic classification, incorporating all sorts of bacteria that could do very different tasks. If they were statistically different at the phylum this would suggest that men and women have some enormously different bacterial flora – I would be more likely to believe that result is an error than a valid result. To give you some perspective, humans would be classified as Chordata at the phylum level – which includes species from mammals, fish, reptiles, birds and even some weird little guy called a sea squirt.

The trade off is of course that it can be difficult to precisely identify bacteria genera or species that would allow us to make a more specific statement about bacterial diversity between genders or even individuals.--

-why the males have higher levels of phyla abundance in their sebum, why females have smaller amounts

-why the abundance of proteobacteria is much higher in males

-why do females have greater diversity of phyla in their microbial makeup

-why more Actinobacteria in women than men, but the second most represented phyla in both sexes?

-sample size for both sexes is too small, need to work with a dataset/research with larger sample size


# Sources Cited


